serializer/src/serializer.rs

```diff
@@ -0,0 +1,264 @@
+        let mut buf = slice;
+        Self::deserialize(&mut buf)
+    }
+
+    fn try_to_vec(&self) -> std::io::Result<Vec<u8>> {
+        let mut buf = Vec::new();
+        self.serialize(&mut buf)?;
+        Ok(buf)
+    }
+}
+
+type ResultStatusTag = u32;
+const RESULT_OK: u32 = 0;
+const RESULT_ERR: u32 = 1;
+
+impl<T, E> Serializer for Result<T, E>
+where
+    T: Serializer + Send + Sync + 'static,
+    E: std::fmt::Display + BorshSerialize + BorshDeserialize + Send + Sync + 'static,
+{
+    fn serialize<W: std::io::Write>(&self, writer: &mut W) -> std::io::Result<()> {
+        match self {
+            Ok(t) => {
+                store!(ResultStatusTag, &RESULT_OK, writer)?;
+                t.serialize(writer)?;
+            }
+            Err(e) => {
+                store!(ResultStatusTag, &RESULT_ERR, writer)?;
+                store!(E, e, writer)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn deserialize(buf: &mut &[u8]) -> std::io::Result<Self> {
+        let tag = load!(ResultStatusTag, buf)?;
+        match tag {
+            RESULT_OK => {
+                let t = T::deserialize(buf)?;
+                Ok(Ok(t))
+            }
+            RESULT_ERR => {
+                let e = E::deserialize(buf)?;
+                Ok(Err(e))
+            }
+            _ => Err(std::io::Error::new(
+                std::io::ErrorKind::InvalidData,
+                "invalid Serializer Result tag",
+            )),
+        }
+    }
+}
+
+type OptionStatusTag = u32;
+const OPTION_SOME: u32 = 1;
+const OPTION_NONE: u32 = 0;
+
+impl<T> Serializer for Option<T>
+where
+    T: Serializer + Send + Sync + 'static,
+{
+    fn serialize<W: std::io::Write>(&self, writer: &mut W) -> std::io::Result<()> {
+        match self {
+            Some(t) => {
+                store!(OptionStatusTag, &OPTION_SOME, writer)?;
+                t.serialize(writer)?;
+            }
+            None => {
+                store!(ResultStatusTag, &OPTION_NONE, writer)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn deserialize(buf: &mut &[u8]) -> std::io::Result<Self> {
+        let tag = load!(ResultStatusTag, buf)?;
+        match tag {
+            OPTION_SOME => {
+                let t = T::deserialize(buf)?;
+                Ok(Some(t))
+            }
+            OPTION_NONE => Ok(None),
+            _ => Err(std::io::Error::new(
+                std::io::ErrorKind::InvalidData,
+                "invalid Serializer Result tag",
+            )),
+        }
+    }
+}
+
+impl<K, V> Serializer for std::collections::HashMap<K, V>
+where
+    K: Serializer + Send + Sync + std::hash::Hash + Eq,
+    V: Serializer + Send + Sync,
+{
+    fn serialize<W: std::io::Write>(&self, writer: &mut W) -> std::io::Result<()> {
+        store!(u32, &(self.len() as u32), writer)?;
+
+        for (k, v) in self.iter() {
+            k.serialize(writer)?;
+            v.serialize(writer)?;
+        }
```

Comment:
hashmap has no deterministic order of iteration, possibly same map can be serialiased to different byte representation.